tutorial/practicals/nemo/4_nemo_openacc/solutions/kernels_trans.py

```diff
@@ -0,0 +1,93 @@
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: R. W. Ford and A. R. Porter, STFC Daresbury Lab
+
+'''A transformation script that adds KERNELS regions enclosed within a DATA
+region to the tracer-advection mini-app.  In order to use it you
+must first install PSyclone. See README.md in the top-level psyclone directory.
+
+Once you have psyclone installed, this may be used by doing:
+
+ $ psyclone -api nemo -s ./kernels_trans.py some_source_file.f90
+
+This should produce a lot of output, ending with generated
+Fortran. Note that the Fortran source files provided to PSyclone must
+have already been preprocessed (if required).
+
+'''
+
+from __future__ import print_function
+from psyclone.psyir.nodes import Loop, Assignment
+from psyclone.transformations import ACCKernelsTrans, ACCDataTrans
+
+
+# Get the PSyclone transformations we will use
+ACC_DATA_TRANS = ACCDataTrans()
+ACC_KERNELS_TRANS = ACCKernelsTrans()
+
+
+def trans(psy):
+    '''A PSyclone-script compliant transformation function.
+
+    :param psy: The PSy layer object to apply transformations to.
+    :type psy: :py:class:`psyclone.psyGen.PSy`
+
+    :returns: the transformed PSy layer object.
+    :rtype: :py:class:`psyclone.psyGen.PSy`
+
+    '''
+    # Get the Schedule of the target routine
+    sched = psy.invokes.get('tra_adv').schedule
+
+    # Find the outer, 'iteration' loop
+    tloop = None
+    for node in sched.children:
+        if isinstance(node, Loop) and node.loop_type == "tracers":
+            tloop = node
+            break
+
+    for node in tloop.loop_body.children:
+        # Enclose explicit loops over vertical levels
+        if isinstance(node, Loop) and node.loop_type == "levels":
+            _ = ACC_KERNELS_TRANS.apply([node])
+        # Enclose array assignments (implicit loops)
+        if isinstance(node, Assignment) and node.is_array_range:
+            _ = ACC_KERNELS_TRANS.apply([node])
+
+    # Finally, enclose the whole of the 'iteration' loop within
+    # a data region
+    # TODO #953 remove `[]` from around tloop argument
```

Comment:
This should not be here. Please ignore my other comments on [] as I'll get #953 onto master before this.

Reply:
Annoyingly #953 didn't entirely fix this as both ACCDataTrans and ACCKernelsTrans have their own `validate` methods that assume they're getting a list. I've removed the TODO but will leave the rest as it is as it's only a small point.

Reply:
Yes I just found this while reviewing. I think it needs to be fixed before this tutorial goes on master - perhaps as part of this tutorial?